From 1eb5068540d106c3f161ffcbacf37a8c7b274291 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Mon, 12 Aug 2013 17:23:59 -0700
Subject: [PATCH 261/974] MIPS: Add new function local_flush_icache_all()

... to flush the entire local ICache.  Default implementation is to do
flush_icache_all().  Processor specific implementations will be in
follow on patches.

Signed-off-by: David Daney <david.daney@cavium.com>
[Original patch taken from Cavium SDK 3.1.2-568]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 arch/mips/include/asm/cacheflush.h | 1 +
 arch/mips/mm/cache.c               | 5 ++++-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/arch/mips/include/asm/cacheflush.h b/arch/mips/include/asm/cacheflush.h
index 69468de..5e0bac5 100644
--- a/arch/mips/include/asm/cacheflush.h
+++ b/arch/mips/include/asm/cacheflush.h
@@ -92,6 +92,7 @@ extern void copy_from_user_page(struct vm_area_struct *vma,
 
 extern void (*flush_cache_sigtramp)(unsigned long addr);
 extern void (*flush_icache_all)(void);
+extern void (*local_flush_icache_all)(void);
 extern void (*local_flush_data_cache_page)(void * addr);
 extern void (*flush_data_cache_page)(unsigned long addr);
 
diff --git a/arch/mips/mm/cache.c b/arch/mips/mm/cache.c
index 5aeb3eb..a75530b 100644
--- a/arch/mips/mm/cache.c
+++ b/arch/mips/mm/cache.c
@@ -45,6 +45,8 @@ void (*flush_cache_sigtramp)(unsigned long addr);
 void (*local_flush_data_cache_page)(void * addr);
 void (*flush_data_cache_page)(unsigned long addr);
 void (*flush_icache_all)(void);
+void (*local_flush_icache_all)(void);
+EXPORT_SYMBOL(local_flush_icache_all);
 
 EXPORT_SYMBOL_GPL(local_flush_data_cache_page);
 EXPORT_SYMBOL(flush_data_cache_page);
@@ -215,7 +217,8 @@ void __cpuinit cpu_cache_init(void)
 
 		octeon_cache_init();
 	}
-
+	if (!local_flush_icache_all)
+		local_flush_icache_all = flush_icache_all;
 	setup_protection_map();
 }
 
-- 
2.6.2

